miio/gateway.py

```diff
@@ -906,154 +1175,325 @@ def update(self):
+        """Turn on Aqara Wall Outlet."""
         return self.send_arg("toggle_plug", ["channel_0", "on"]).pop()
 
     @command()
     def off(self):
-        """Turn off Aqara Wall Outlet"""
+        """Turn off Aqara Wall Outlet."""
         return self.send_arg("toggle_plug", ["channel_0", "off"]).pop()
 
 
-class Cube(SubDevice):
-    """Subdevice Cube specific properties and methods"""
+class AqaraSmartBulbE27(SubDevice):
+    """Subdevice AqaraSmartBulbE27 specific properties and methods."""
 
     properties = []
-    _model = "lumi.sensor_cube.v1"
+    _zigbee_model = "lumi.light.aqcn02"
+    _model = "ZNLDP12LM"
+    _name = "Smart bulb E27"
 
 
-class AqaraSquareButton(SubDevice):
-    """Subdevice AqaraSquareButton specific properties and methods"""
+class CubeV2(SubDevice):
+    """Subdevice CubeV2 specific properties and methods."""
 
     properties = []
-    _model = "lumi.sensor_switch.aq3"
+    _zigbee_model = "lumi.sensor_cube.aqgl01"
+    _model = "MFKZQ01LM"
+    _name = "Cube"
 
 
-class SwitchOneChannel(SubDevice):
-    """Subdevice SwitchOneChannel specific properties and methods"""
+class LockS2(SubDevice):
+    """Subdevice LockS2 specific properties and methods."""
 
-    properties = ["neutral_0"]
-    _model = "lumi.ctrl_neutral1.v1"
+    properties = []
+    _zigbee_model = "lumi.lock.acn02"
+    _model = "ZNMS12LM"
+    _name = "Door lock S2"
 
-    @attr.s(auto_attribs=True)
-    class props:
-        """Device specific properties"""
 
-        status: str = None  # 'on' / 'off'
+class Curtain(SubDevice):
+    """Subdevice Curtain specific properties and methods."""
 
-    @command()
-    def update(self):
-        """Update all device properties"""
-        values = self.get_property_exp(self.properties)
-        self._props.status = values[0]
+    properties = []
+    _zigbee_model = "lumi.curtain.aq2"
+    _model = "ZNGZDJ11LM"
+    _name = "Curtain"
 
-    @command()
-    def toggle(self):
-        """Toggle Switch One Channel"""
-        return self.send_arg("toggle_ctrl_neutral", ["channel_0", "toggle"]).pop()
 
-    @command()
-    def on(self):
-        """Turn on Switch One Channel"""
-        return self.send_arg("toggle_ctrl_neutral", ["channel_0", "on"]).pop()
+class CurtainB1(SubDevice):
+    """Subdevice CurtainB1 specific properties and methods."""
 
-    @command()
-    def off(self):
-        """Turn off Switch One Channel"""
-        return self.send_arg("toggle_ctrl_neutral", ["channel_0", "off"]).pop()
+    properties = []
+    _zigbee_model = "lumi.curtain.hagl04"
+    _model = "ZNCLDJ12LM"
+    _name = "Curtain B1"
 
 
-class RemoteSwitchSingleV1(SubDevice):
-    """Subdevice RemoteSwitchSingleV1 specific properties and methods"""
+class LockV1(SubDevice):
+    """Subdevice LockV1 specific properties and methods."""
+
+    properties = []
+    _zigbee_model = "lumi.lock.v1"
+    _model = "A6121"
+    _name = "Vima cylinder lock"
+
+
+class IkeaBulb82(SubDevice):
```

Comment:
Yea the gateway.py file grew a lot and is getting way too big.
Was already thinking the same thing, at least all subdevices schould be moved to a gateway_subdevices.py file or something....

for the IkeaBulbs and simular classes, I think it would be best to define a new class `IkeaBulb` that inherits properties from `SubDevice` class, then all individual model classes like `IkeaBulb82` can inherit all common functions from the `IkeaBulb` class.

Not sure if that is what you were trying to say....

Reply:
Maybe split up all subdevices in files which have one common class that they inherit from. so all IkeaBulbs in one file, all cubes in one, all buttons in one etc.

Reply:
Yes, having a common base class would make it much cleaner already. I don't think having a `gateway_subdevices.py` is a good idea, it's much better to have a structure like:
```
miio/gateway/
             gateway.py
             gateway_zigbee.py
             gateway_light.py
             gateway_radio.py
```
and so on. To extend that example, here's another idea that would make it much nicer:
```
miio/gateway/
            gateway.py
            subdevice.py
miio/gateway/aqara/
                  cube.py
                  switch.py
miio/gateway/ikea/
                  bulb.py
```